### PR TITLE
add sitemap functionality

### LINKED ISSUE #798: Optimise Google Scholar indexation
**Current behaviour**

Our service [RERO DOC](http://doc.rero.ch/), well indexed by Google Scholar, is progressively replaced by [SONAR](https://sonar.ch/), which is used by various universities in Switzerland.
Some documents in SONAR are not indexed by Google Scholar, example: https://sonar.ch/global/documents/312841

Diagnose from Google:
* there is no way for the crawler to get a list of all the individual records

**Wanted behaviour**

All documents are indexed in Google Scholar

**To do**

1. Add a sitemap that is updated daily or weekly and the url for that listed in the robots.txt.   Then contact back Google to configure the crawler to use it to get the list of document urls. If individual files are larger than 5MB, it would be best to split the files with a max size of 5MB and use a sitemap index file to list the sitemaps. You can see an example at https://hal.inria.fr/robots which lists sitemap: https://hal.inria.fr/robots/sitemap
    * To clarify: for dedicated repo, present the dedicated URL or the global URL? Both?
2. display the full abstract for users arrive from Google Scholar & Google web search (no "Show more" option)

**Remarks**

Maybe look at this: https://github.com/inveniosoftware/flask-sitemap

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Adds a new field serverName on organisation resource.
* Implements sitemap generation for global and dedicated organisations.
* Implements generation of the file robot.txt dynamically.
* Closes #798.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

⚠️ **ES update mapping**
⚠️ **Production**
* make the flag **True** on INVENIO_SONAR_APP_PRODUCTION_STATE
* set  INVENIO_SONAR_APP_SITEMAP_FOLDER_PATH

## How to test?

- `poetry shell` 
- `./script/bootstrap`
- `./script/setup --clear-data`

### Sitemap
- `invenio oaiharvester harvest -n rerodoc -q -a max=200 -f 2021-10-01`
- `invenio sitemap generate -d sonar.rero.ch` (for global view)
- `invenio sitemap generate -d folia.unifr.ch` (for global view)
- Global sitemap: `https://localhost:5000/global/sitemap.xml`
- Organisation sitemap: `https://localhost:5000/unifr/sitemap.xml` (if you are some documents)

### Robot.txt
by default: disallow indexation
- `https://localhost:5000/robot.txt`
- Change constant `SONAR_APP_PRODUCTION_STATE` to True
- `https://localhost:5000/robot.txt` (allow indexation and show sitemap link)

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: sonar/modules/organisations/jsonschemas/organisations/organisation-v1.0.0_src.json [hunk 1]
```diff
@@ -146,6 +146,29 @@
...
       }
     },
+    "serverName": {
+      "title": "Server name (without http)",
+      "description": "Organisation server name for dedicated.",
+      "type": "string",
+      "form": {
+        "hideExpression": "!field.model.isDedicated",
+        "expressionProperties": {
+          "templateOptions.required": "!field.model.isDedicated"
```
Comment (PascalRepond) [inline sonar/modules/organisations/jsonschemas/organisations/organisation-v1.0.0_src.json:156]: Remove `!` ? 
Field is required if organisation is dedicated.

File: sonar/modules/organisations/api.py [hunk 2]
```diff
@@ -84,6 +84,19 @@ def get_shared_or_dedicated_list(self):
                                ['pid', 'name', 'isShared',
                                 'isDedicated']).execute().hits
 
+    def get_organisation_pid_by_server_name(self, server_name):
+        """Get organisation by server_name."""
```
Comment (BadrAly) [inline sonar/modules/organisations/api.py:88]: add param and return description?

File: sonar/modules/sitemap/cli/sitemap.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from sonar.modules.sitemap.sitemap import sitemap_generate
+
+
+@click.group()
+def sitemap():
+    """Sitemap."""
+
+@sitemap.command()
+@click.option(
+    '-d', '--server-name', 'server_name', required=True, default=None)
```
Comment (BadrAly) [inline sonar/modules/sitemap/cli/sitemap.py:34]: `-s` instead of `-d`?

**Review by vgranata — APPROVED** [review]

**Comment by pronguen** [comment]
- [ ] NPR/18.08: The file should be named `robots.txt` instead of `robot.txt`

**Pull request merged**